File: scratch/graph950/two_tier_orchestrator.py
import time, sys, os, argparse, collections
from typing import Dict, List, Set, Tuple, Any, Optional

from scratch.graph950.two_tier_decomposer import load_graph, decompose_graph
from scratch.graph950.pinpointed_strip_solver import PinpointedStripSolver
from scratch.graph950.global_demand_coordinator import GlobalDemandCoordinator
from scratch.graph950.macro_splicer import splice_macro_tour, verify_tour_on_raw_graph

def write_hcp_tour(tour: List[int], output_path: str):
    """
    Writes a certified tour in standard TSPLIB/HCP format.
    """
    os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
    with open(output_path, 'w') as f:
        f.write("NAME : graph950.hcp.tour\n")
        f.write("TYPE : TOUR\n")
        f.write(f"DIMENSION : {len(tour)}\n")
        f.write("TOUR_SECTION\n")
        for v in tour:
            f.write(f"{v}\n")
        f.write("-1\n")
        f.write("EOF\n")

def solve_graph950_two_tier(
    graph_path: str = "FHCPCS-col/graph950.col",
    timeout: float = 1800.0,
    dry_run: bool = False,
    output_path: str = "scratch/graph950/found_tour_puresat.hcp"
) -> bool:
    t_start = time.time()
    print(f"=== Starting Two-Tier Demand-Coordinated Solver on {graph_path} ===")

    if not os.path.exists(graph_path):
        print(f"Error: Target graph file '{graph_path}' not found.")
        return False

    G, degs = load_graph(graph_path)
    decomp = decompose_graph(G, degs)
    print(
        f"Decomposition: {len(decomp.all_hubs)} hubs "
        f"({len(decomp.s_hubs)} S, {len(decomp.b_hubs)} B, {len(decomp.m_hubs)} M), "
        f"{len(decomp.strips)} strips ({len([s for s in decomp.strips if len(s) == 125])} large)"
    )

    if dry_run:
        print("[DRY RUN] Initialization complete. Exiting without solving.")
        return True

    strip_solver = PinpointedStripSolver(G, decomp)
    coordinator = GlobalDemandCoordinator(G, decomp)

    outer_it = 0
    while True:
        outer_it += 1
        elapsed = time.time() - t_start
        if elapsed > timeout:
            print(f"[TIMEOUT] Reached global {timeout}s limit at iteration {outer_it}")
            return False

        if outer_it % 10 == 1 or outer_it <= 5:
            print(f"\n--- Outer Iteration {outer_it} ({elapsed:.1f}s) ---")
            
        is_sat, hh_edges, strip_demands = coordinator.solve_assignment()
        if not is_sat:
            print("Coordinator returned UNSAT: search space exhausted.")
            return False

        if outer_it % 10 == 1 or outer_it <= 5:
            print(f"Coordinator assigned {len(hh_edges)} Hub-Hub edges across {len(decomp.strips)} strips")

        all_strips_sat = True
        strip_paths = {}

        for si, s in enumerate(decomp.strips):
            if time.time() - t_start > timeout:
                print(f"[TIMEOUT] Global timeout reached during strip solving.")
                return False

            s_hub = list(decomp.strip_adj_hubs[si] & set(decomp.s_hubs))[0] if (decomp.strip_adj_hubs[si] & set(decomp.s_hubs)) else None
            b_hub = list(decomp.strip_adj_hubs[si] & set(decomp.b_hubs))[0] if (decomp.strip_adj_hubs[si] & set(decomp.b_hubs)) else None
            dem = strip_demands.get(si, {})

            tot_d = sum(dem.values())
            K = tot_d // 2 if tot_d >= 2 else (1 if len(s) < 10 else 4)

            sat, res = strip_solver.solve_strip(si, dem, s_hub, b_hub, K=K)

            if not sat:
                all_strips_sat = False
                failed_core = res if isinstance(res, list) else list(dem.keys())
                coordinator.add_conflict_clause(si, dem, failed_core)
                if outer_it % 10 == 1 or outer_it <= 5:
                    print(f"  Strip {si:2d} ({len(s)}v) UNSAT with core {failed_core} -> conflict learned")
                break
            else:
                strip_paths[si] = res

        if not all_strips_sat:
            continue

        print(f"All {len(decomp.strips)} strips SATISFIED at iter {outer_it} ({time.time()-t_start:.1f}s)! Splicing full tour...")
        is_valid, res = splice_macro_tour(G, decomp, hh_edges, strip_paths, strip_demands)

        if is_valid:
            tour = res
            print(f"SUCCESS! Single Hamiltonian tour formed with {len(tour)} vertices!")
            if verify_tour_on_raw_graph(tour, G):
                print(f"CERTIFICATION PASSED: Verified tour independently on raw graph G!")
                write_hcp_tour(tour, output_path)
                print(f"Wrote certified tour to {output_path}")
                print(f"Total time elapsed: {time.time() - t_start:.2f}s")
                return True
            else:
                print("Verification on raw graph failed.")
                return False
        else:
            subtours = res
            if isinstance(subtours, list) and len(subtours) > 0:
                print(f"Splicer detected {len(subtours)} disconnected subtours -> adding macro cut clauses")
                for cyc in subtours:
                    coordinator.add_macro_cut(set(cyc), hh_edges, strip_demands)

def main():
    parser = argparse.ArgumentParser(description="End-to-End Two-Tier Demand Coordinator Solver")
    parser.add_argument("--graph", type=str, default="FHCPCS-col/graph950.col", help="Path to .col graph file")
    parser.add_argument("--timeout", type=float, default=1800.0, help="Wall-clock timeout in seconds")
    parser.add_argument("--out", type=str, default="scratch/graph950/found_tour_puresat.hcp", help="Output path for .hcp tour")
    parser.add_argument("--dry_run", action="store_true", help="Perform dry run initialization without solving")
    args = parser.parse_args()

    success = solve_graph950_two_tier(
        graph_path=args.graph,
        timeout=args.timeout,
        dry_run=args.dry_run,
        output_path=args.out
    )
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()

File: src/cegar-fix/src/two_tier_orchestrator.rs
use crate::component_meta_graph::ComponentMetaGraph;
use crate::global_demand_coordinator::GlobalDemandCoordinator;
use crate::graph::Graph;
use crate::macro_splicer::{splice_macro_tour, verify_tour_on_raw_graph};
use crate::pinpointed_strip_solver::PinpointedStripSolver;
use crate::two_tier_decomposer::decompose_graph;
use std::collections::{HashMap, HashSet};
use std::path::Path;
use std::time::Instant;

pub struct TwoTierSolverOptions {
    pub timeout_secs: f64,
    pub max_iterations: usize,
    pub enable_patching: bool,
    pub output_path: Option<String>,
}

impl Default for TwoTierSolverOptions {
    fn default() -> Self {
        Self {
            timeout_secs: 1800.0,
            max_iterations: 50_000,
            enable_patching: true,
            output_path: None,
        }
    }
}

#[derive(Debug, Clone)]
pub struct TwoTierOptions {
    pub timeout_secs: f64,
    pub output_tour: Option<String>,
}

impl Default for TwoTierOptions {
    fn default() -> Self {
        Self {
            timeout_secs: 1800.0,
            output_tour: None,
        }
    }
}

pub struct TwoTierOrchestrator;

impl TwoTierOrchestrator {
    pub fn solve(g: &Graph, options: &TwoTierOptions) -> Option<Vec<i32>> {
        let opt = TwoTierSolverOptions {
            timeout_secs: options.timeout_secs,
            max_iterations: 50_000,
            enable_patching: true,
            output_path: options.output_tour.clone(),
        };
        solve_graph_two_tier(g, &opt)
    }
}

/// Writes a certified tour in standard TSPLIB/HCP format.
pub fn write_hcp_tour(tour: &[i32], output_path: &str) -> std::io::Result<()> {
    let name = Path::new(output_path)
        .file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("tour");
    crate::tour_verifier::TourVerifier::write_tsplib_hcp(tour, name, output_path)
}

/// Top-level solver function integrating:
/// 1. decompose_graph(&g)
/// 2. GlobalDemandCoordinator
/// 3. PinpointedStripSolver
/// 4. splice_macro_tour + patch_cycles_2opt
/// 5. Indicator Cut-Crossing SECs on subtours
/// 6. Independent tour certification & output to HCP format
pub fn solve_graph_two_tier(g: &Graph, options: &TwoTierSolverOptions) -> Option<Vec<i32>> {
    let start_time = Instant::now();
    println!("=== Starting Two-Tier Demand-Coordinated Solver in Rust ===");

    let decomp = decompose_graph(g);
    let large_strips_count = decomp.strips.iter().filter(|s| s.len() >= 10).count();
    println!(
        "Decomposition: {} hubs ({} S, {} B, {} M), {} strips ({} large)",
        decomp.all_hubs.len(),
        decomp.s_hubs.len(),
        decomp.b_hubs.len(),
        decomp.m_hubs.len(),
        decomp.strips.len(),
        large_strips_count
    );

    let enable_mtz = decomp.all_hubs.len() >= 2 && decomp.all_hubs.len() <= 250;
    if enable_mtz {
        println!(
            "Active Macro Order-Encoding (MTZ) enabled on {} hubs",
            decomp.all_hubs.len()
        );
    }

    let mut coordinator = GlobalDemandCoordinator::new_with_mtz(g, &decomp, enable_mtz);
    let mut strip_solver = PinpointedStripSolver::new(g, &decomp);
    let mut strip_cache: HashMap<(usize, Vec<(i32, usize)>, usize), Result<Vec<Vec<i32>>, Vec<i32>>> = HashMap::new();

    let mut outer_it = 0;
    while outer_it < options.max_iterations {
        outer_it += 1;
        let elapsed = start_time.elapsed().as_secs_f64();
        if elapsed > options.timeout_secs {
            println!(
                "[TIMEOUT] Reached global {:.1}s limit at iteration {}",
                options.timeout_secs, outer_it
            );
            return None;
        }

        if outer_it % 10 == 1 || outer_it <= 5 {
            println!("\n--- Outer Iteration {} ({:.1}s) ---", outer_it, elapsed);
        }

        let assignment = coordinator.solve_assignment();
        if assignment.is_none() {
            println!("Coordinator returned UNSAT: search space exhausted.");
            return None;
        }

        let (hh_edges, strip_demands) = assignment.unwrap();
        if outer_it % 10 == 1 || outer_it <= 5 {
            println!(
                "Coordinator assigned {} Hub-Hub edges across {} strips",
                hh_edges.len(),
                decomp.strips.len()
            );
        }

        let mut all_strips_sat = true;
        let mut strip_paths = HashMap::new();

        for (si, s) in decomp.strips.iter().enumerate() {
            if start_time.elapsed().as_secs_f64() > options.timeout_secs {
                println!("[TIMEOUT] Global timeout reached during strip solving.");
                return None;
            }

            let dem = strip_demands.get(&si).cloned().unwrap_or_default();
            let tot_d: usize = dem.values().sum();
            let k = if tot_d >= 2 {
                tot_d / 2
            } else if s.len() < 10 {
                1
            } else {
                4
            };

            let s_hub = decomp.strip_adj_hubs.get(&si).and_then(|adj| {
                decomp.s_hubs.iter().find(|h| adj.contains(h)).copied()
            });
            let b_hub = decomp.strip_adj_hubs.get(&si).and_then(|adj| {
                decomp.b_hubs.iter().find(|h| adj.contains(h)).copied()
            });

            let mut dem_vec: Vec<(i32, usize)> = dem.iter().map(|(&h, &d)| (h, d)).filter(|&(_, d)| d > 0).collect();
            dem_vec.sort_unstable();
            let cache_key = (si, dem_vec, k);

            let res = if let Some(cached) = strip_cache.get(&cache_key) {
                cached.clone()
            } else {
                let r = strip_solver.solve_strip(si, &dem, s_hub, b_hub, k);
                strip_cache.insert(cache_key, r.clone());
                r
            };

            match res {
                Ok(paths) => {
                    strip_paths.insert(si, paths);
                }
                Err(failed_core) => {
                    all_strips_sat = false;
                    coordinator.add_conflict_clause(si, &dem, &failed_core);
                    if outer_it % 10 == 1 || outer_it <= 5 {
                        println!(
                            "  Strip {:2} ({}v) UNSAT with core {:?} -> conflict learned",
                            si,
                            s.len(),
                            failed_core
                        );
                    }
                    break;
                }
            }
        }

        if !all_strips_sat {
            continue;
        }

        println!(
            "All {} strips SATISFIED at iter {} ({:.1}s)! Splicing full tour...",
            decomp.strips.len(),
            outer_it,
            start_time.elapsed().as_secs_f64()
        );

        let (is_single, cycles) = splice_macro_tour(
            g,
            &decomp,
            &hh_edges,
            &strip_paths,
            &strip_demands,
            options.enable_patching,
        );

        if is_single && cycles.len() == 1 {
            let tour = &cycles[0];
            println!(
                "SUCCESS! Single Hamiltonian tour formed with {} vertices!",
                tour.len()
            );
            if verify_tour_on_raw_graph(tour, g) {
                println!("CERTIFICATION PASSED: Verified tour independently on raw graph G!");
                let out_path = options
                    .output_path
                    .as_deref()
                    .unwrap_or("scratch/graph950/found_tour_rust.hcp");
                if let Err(e) = write_hcp_tour(tour, out_path) {
                    eprintln!("Warning: failed to write tour to {}: {}", out_path, e);
                } else {
                    println!("Wrote certified tour to {}", out_path);
                }
                println!(
                    "Total time elapsed: {:.2}s",
                    start_time.elapsed().as_secs_f64()
                );
                return Some(tour.clone());
            } else {
                eprintln!("Verification on raw graph failed.");
                return None;
            }
        } else {
            if !cycles.is_empty() {
                println!(
                    "Splicer detected {} disconnected subtours -> adding macro cut clauses",
                    cycles.len()
                );
                if cycles.len() > 1 {
                    let meta_graph = ComponentMetaGraph::build(&cycles, g);
                    if meta_graph.meta_components.len() > 1 {
                        println!(
                            "Meta-graph partitioned into {} disconnected components -> generating multi-component SEC cuts",
                            meta_graph.meta_components.len()
                        );
                        coordinator.add_meta_component_cuts(meta_graph.get_meta_components(), &cycles);
                    }
                }
                coordinator.add_exact_subtour_block(&hh_edges, &strip_demands, &cycles);
                for cyc in &cycles {
                    let cyc_verts: HashSet<i32> = cyc.iter().copied().collect();
                    coordinator.add_macro_cut(&cyc_verts);
                    coordinator.inject_component_mtz(cyc);
                }
            } else {
                println!("Splicer failed boundary matching -> learning conflict clauses on strips");
                for (si, _) in decomp.strips.iter().enumerate() {
                    if let Some(dem) = strip_demands.get(&si) {
                        let failed_hubs: Vec<i32> = dem.keys().copied().collect();
                        coordinator.add_conflict_clause(si, dem, &failed_hubs);
                    }
                }
            }
        }
    }

    None
}
